parlai/scripts/convert_fairseq_to_parlai.py

```diff
@@ -0,0 +1,485 @@
+                            mapped_key.replace(
+                                attention_project_name, 'k_lin.{}'.format(weightorbias)
+                            )
+                        ] = weights[1]
+                        return_dict[
+                            mapped_key.replace(
+                                attention_project_name, 'v_lin.{}'.format(weightorbias)
+                            )
+                        ] = weights[2]
+                continue
+            elif (
+                'v_proj' in mapped_key
+                or 'k_proj' in mapped_key
+                or 'q_proj' in mapped_key
+            ):
+                mapped_key = mapped_key.replace('v_proj', 'v_lin')
+                mapped_key = mapped_key.replace('q_proj', 'q_lin')
+                mapped_key = mapped_key.replace('k_proj', 'k_lin')
+
+            # 4. Replace FFN layers
+            mapped_key = mapped_key.replace('fc1', 'ffn.lin1')
+            mapped_key = mapped_key.replace('fc2', 'ffn.lin2')
+            mapped_key = mapped_key.replace('layer_norms.0', 'norm1')
+            mapped_key = mapped_key.replace('layer_norms.1', 'norm2')
+            mapped_key = mapped_key.replace('out_proj', 'out_lin')
+
+            # 5. Fix layer norms
+            if 'encoder.' in mapped_key:
+                mapped_key = mapped_key.replace('attention_layer_norm', 'norm1')
+                mapped_key = mapped_key.replace('final_layer_norm', 'norm2')
+            else:
+                mapped_key = mapped_key.replace('self_attention_layer_norm', 'norm1')
+                mapped_key = mapped_key.replace('encoder_attention_layer_norm', 'norm2')
+                mapped_key = mapped_key.replace('final_layer_norm', 'norm3')
+
+            mapped_key = mapped_key.replace(
+                'encoder.layer_norm', 'encoder.norm_embeddings'
+            )
+            mapped_key = mapped_key.replace(
+                'encoder.layernorm_embedding', 'encoder.norm_embeddings'
+            )
+            mapped_key = mapped_key.replace(
+                'decoder.layer_norm', 'decoder.norm_embeddings'
+            )
+            mapped_key = mapped_key.replace(
+                'decoder.layernorm_embedding', 'decoder.norm_embeddings'
+            )
+
+            weight = state_dict[each_key]
+            return_dict[mapped_key] = weight
+
+        # 6. Shuffle embedding matrix given dictionary.
+        enc_emb_key = 'encoder.embeddings.weight'
+        bart_dict = os.path.join(
+            opt['datapath'], 'models/bart_models/bart.large/dict.txt'
+        )
+        with open(bart_dict) as f:
+            offset_dict = {i: l.split()[0] for i, l in enumerate(f.readlines())}
+        new_embs = return_dict[enc_emb_key].clone()
+        for idx, new_idx in offset_dict.items():
+            try:
+                new_embs[int(new_idx) + 4] = return_dict[enc_emb_key][idx + 4]
+            except ValueError:
+                # if idx is not an int
```

Comment:
😮  when does this happen?

Reply:
the fairseq dictionary is a mapping from tok index to "word", however the "words" are just string representations of the token values, so like `'504': 106` might be a value in there.

EXCEPT for special tokens (pad, bos, eos, unk) and fp16 dict-pad tokens (madeupword0001, madeupword0002, etc.), which show up as actual strings. So sometimes the `int(new_idx)` conversion fails